src/Rules/LibIME.hs

```diff
@@ -0,0 +1,144 @@
+module Rules.LibIME
+  (
+    libIMERule,
+    chineseAddonsRule,
+  )
+where
+
+import Base
+import Data.Maybe (fromJust)
+
+fcitxDataUrl :: String
+fcitxDataUrl = "https://download.fcitx-im.org/data/"
+
+dictNames :: [String]
+dictNames = ["sc", "extb"]
+
+tableDictNames :: [String]
+tableDictNames = ["db", "erbi", "qxm", "wanfeng", "wbpy", "wbx", "zrm", "cj"]
+
+libIMERule :: Rules ()
+libIMERule = do
+  libIMEToolsRule
+  lmRule
+  pinyinDictRule
+  tableDictRule
+  "libime" ~> do
+    copyFile' (outputDir </> "sc.dict") $ outputDir </> "libime" </> "data" </> "sc.dict"
+    copyFile' (outputDir </> "extb.dict") $ outputDir </> "libime" </> "data" </> "extb.dict"
+    copyFile' (outputDir </> "sc.lm") $ outputDir </> "libime" </> "data" </> "zh_CN.lm"
+    copyFile' (outputDir </> "sc.lm.predict") $ outputDir </> "libime" </> "data" </> "zh_CN.lm.predict"
+    forM_ tableDictNames $ \table ->
+      let name = table <.> "main.dict"
+       in copyFile' (outputDir </> name) (outputDir </> "libime" </> "table" </> name)
+
+libIMEToolsRule :: Rules ()
+libIMEToolsRule = do
+  "libime-tools" ~> do
+    let libIMESrc = "libime"
+    cmd_ "cmake" "-B" (libIMESrc </> "build") "-G" "Ninja" 
+      [
+        "-DCMAKE_BUILD_TYPE=Release",
+        "-DCMAKE_INSTALL_PREFIX=" <> outputDir,
+        "-DENABLE_TEST=OFF"
+      ]
+      libIMESrc
+    cmd_ "cmake" "--build" (libIMESrc </> "build") "--target"
+      [
+        "libime_slm_build_binary",
+        "libime_prediction",
+        "libime_pinyindict",
+        "libime_history",
+        "libime_tabledict"
+      ]
+    -- ignore install errors
+    Exit c <- cmd "cmake" "--install" (libIMESrc </> "build") "--component" "lib"
+    Exit c <- cmd "cmake" "--install" (libIMESrc </> "build") "--component" "header"
+    Exit c <- cmd "cmake" "--install" (libIMESrc </> "build") "--component" "tools"
+    Exit c <- cmd "cmake" "--install" (libIMESrc </> "build") "--component" "Devel"
+    cmd_ "echo" ("installed to " <> outputDir)
+
+lmRule :: Rules ()
+lmRule = do
+  outputDir </> "lm_sc.arpa" %> \out -> do
+    src <- getConfig' "lm_sc"
+    sha256 <- getConfig' "lm_sc_sha256"
+    tar <- download fcitxDataUrl src sha256
+    cmd_ "tar" "xf" tar "-C" outputDir (takeFileName out)
+  outputDir </> "sc.lm" %> \out -> do
+    let src = outputDir </> "lm_sc.arpa"
+    need ["libime-tools", src]
+    cmd_ (AddEnv "LD_LIBRARY_PATH" (outputDir </> "lib")) (outputDir </> "bin" </> "libime_slm_build_binary") "-s -a 22 -q 8 trie" src out
```

Comment:
It would be great if we can factor out a function to call libime executables. I'll take care of this.